pkg/middleware/stored_session.go

```diff
@@ -108,49 +109,56 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
 
 // refreshSessionIfNeeded will attempt to refresh a session if the session
 // is older than the refresh period.
-// It is assumed that if the provider refreshes the session, the session is now
-// valid.
-// If the session requires refreshing but the provider does not refresh it,
-// we must validate the session to ensure that the returned session is still
-// valid.
+// Success or fail, we will then validate the session.
 func (s *storedSessionLoader) refreshSessionIfNeeded(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
 	if s.refreshPeriod <= time.Duration(0) || session.Age() < s.refreshPeriod {
 		// Refresh is disabled or the session is not old enough, do nothing
 		return nil
 	}
 
-	logger.Printf("Refreshing %s old session cookie for %s (refresh after %s)", session.Age(), session, s.refreshPeriod)
-	refreshed, err := s.refreshSessionWithProvider(rw, req, session)
+	logger.Printf("Refreshing session - User: %s; SessionAge: %s", session.User, session.Age())
+	err := s.refreshSession(rw, req, session)
 	if err != nil {
-		return err
+		// If a preemptive refresh fails, we still keep the session
+		// if validateSession succeeds.
+		logger.Errorf("Unable to refresh session: %v", err)
 	}
 
-	if !refreshed {
-		// Session wasn't refreshed, so make sure it's still valid
-		return s.validateSession(req.Context(), session)
-	}
-	return nil
+	// Validate all sessions after any Redeem/Refresh operation (fail or success)
+	return s.validateSession(req.Context(), session)
 }
 
-// refreshSessionWithProvider attempts to refresh the sessinon with the provider
+// refreshSession attempts to refresh the session with the provider
 // and will save the session if it was updated.
-func (s *storedSessionLoader) refreshSessionWithProvider(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) (bool, error) {
-	refreshed, err := s.refreshSessionWithProviderIfNeeded(req.Context(), session)
-	if err != nil {
-		return false, fmt.Errorf("error refreshing access token: %v", err)
+func (s *storedSessionLoader) refreshSession(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
+	refreshed, err := s.sessionRefresher(req.Context(), session)
+	// HACK:
+	// Providers that don't implement `RefreshSession` use the default
+	// implementation which returns `ErrNotImplemented`.
+	// Fallthrough & update `session.CreatedAt` to reset the refresh timer so
+	// that `ValidateSession` isn't triggered every subsequent request and is
+	// only called once during this request.
+	if !errors.Is(err, providers.ErrNotImplemented) {
+		if err != nil {
+			return fmt.Errorf("error refreshing tokens: %v", err)
+		}
+		// Session not refreshed, nothing to persist.
+		if !refreshed {
+			return nil
+		}
```

Comment:
For some reason I'm really struggling to follow how and why this is correct :thinking: 

I'm wondering if there is a different hacky way to make this easier to understand

Is this the same as 
```
if err !=nil && !errors.Is(err, providers.ErrNotImplemented) {
   return err
}

if errors.Is(err, providers.ErrNotImplemented) {
    // Insert comment about this being a hack here
    refreshed = true
}

if !refreshed {
    return nil
}
```

Reply:
Yeah - this looks like it unnests it well to make it clearer. (I originally had something similar, but consolidated the `ErrNotImplemented` check since it was happening multiple times BUT at the expense of more nesting).

This is getting back to the original design though -- just having the default implementation return `true` for `refreshed` even though it didn't do anything.

Want to leave that as `false` in the implementation but override here at the source of the need for the hack with the explanation?

Reply:
> Want to leave that as false in the implementation but override here at the source of the need for the hack with the explanation?

I think this makes the most sense, it's up to the caller to determine how they handle a not implemented in most cases I see. So i think it makes sense to have the hack description right in the logic here, rather than in the provider implementation where we might not see/understand it as we don't have the context